File: go/src/reader/reader.go
package reader

import (
	"errors"
	"regexp"
	"strconv"
	"strings"
	//"fmt"
)

import (
	. "types"
)

type Reader interface {
	next() *string
	peek() *string
}

type TokenReader struct {
	tokens   []string
	position int
}

func (tr *TokenReader) next() *string {
	if tr.position >= len(tr.tokens) {
		return nil
	}
	token := tr.tokens[tr.position]
	tr.position = tr.position + 1
	return &token
}

func (tr *TokenReader) peek() *string {
	if tr.position >= len(tr.tokens) {
		return nil
	}
	return &tr.tokens[tr.position]
}

func tokenize(str string) []string {
	results := make([]string, 0, 1)
	// Work around lack of quoting in backtick
	re := regexp.MustCompile(`[\s,]*(~@|[\[\]{}()'` + "`" +
		`~^@]|"(?:\\.|[^\\"])*"|;.*|[^\s\[\]{}('"` + "`" +
		`,;)]*)`)
	for _, group := range re.FindAllStringSubmatch(str, -1) {
		if (group[1] == "") || (group[1][0] == ';') {
			continue
		}
		results = append(results, group[1])
	}
	return results
}

func read_atom(rdr Reader) (MalType, error) {
	token := rdr.next()
	if token == nil {
		return nil, errors.New("read_atom underflow")
	}
	if match, _ := regexp.MatchString(`^-?[0-9]+$`, *token); match {
		var i int
		var e error
		if i, e = strconv.Atoi(*token); e != nil {
			return nil, errors.New("number parse error")
		}
		return i, nil
	} else if (*token)[0] == '"' {
		str := (*token)[1 : len(*token)-1]
		return strings.Replace(
			strings.Replace(str, `\"`, `"`, -1),
			`\n`, "\n", -1), nil
	} else if (*token)[0] == ':' {
		return NewKeyword((*token)[1:len(*token)])
	} else if *token == "nil" {
		return nil, nil
	} else if *token == "true" {
		return true, nil
	} else if *token == "false" {
		return false, nil
	} else {
		return Symbol{*token}, nil
	}
	return token, nil
}

func read_list(rdr Reader, start string, end string) (MalType, error) {
	token := rdr.next()
	if token == nil {
		return nil, errors.New("read_list underflow")
	}
	if *token != start {
		return nil, errors.New("expected '" + start + "'")
	}

	ast_list := []MalType{}
	token = rdr.peek()
	for ; true; token = rdr.peek() {
		if token == nil {
			return nil, errors.New("exepected '" + end + "', got EOF")
		}
		if *token == end {
			break
		}
		f, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		ast_list = append(ast_list, f)
	}
	rdr.next()
	return List{ast_list, nil}, nil
}

func read_vector(rdr Reader) (MalType, error) {
	lst, e := read_list(rdr, "[", "]")
	if e != nil {
		return nil, e
	}
	vec := Vector{lst.(List).Val, nil}
	return vec, nil
}

func read_hash_map(rdr Reader) (MalType, error) {
	mal_lst, e := read_list(rdr, "{", "}")
	if e != nil {
		return nil, e
	}
	return NewHashMap(mal_lst)
}

func read_form(rdr Reader) (MalType, error) {
	token := rdr.peek()
	if token == nil {
		return nil, errors.New("read_form underflow")
	}
	switch *token {

	case `'`:
		rdr.next()
		form, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		return List{[]MalType{Symbol{"quote"}, form}, nil}, nil
	case "`":
		rdr.next()
		form, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		return List{[]MalType{Symbol{"quasiquote"}, form}, nil}, nil
	case `~`:
		rdr.next()
		form, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		return List{[]MalType{Symbol{"unquote"}, form}, nil}, nil
	case `~@`:
		rdr.next()
		form, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		return List{[]MalType{Symbol{"splice-unquote"}, form}, nil}, nil
	case `^`:
		rdr.next()
		meta, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		form, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		return List{[]MalType{Symbol{"with-meta"}, form, meta}, nil}, nil
	case `@`:
		rdr.next()
		form, e := read_form(rdr)
		if e != nil {
			return nil, e
		}
		return List{[]MalType{Symbol{"deref"}, form}, nil}, nil

	// list
	case ")":
		return nil, errors.New("unexpected ')'")
	case "(":
		return read_list(rdr, "(", ")")

	// vector
	case "]":
		return nil, errors.New("unexpected ']'")
	case "[":
		return read_vector(rdr)

	// hash-map
	case "}":
		return nil, errors.New("unexpected '}'")
	case "{":
		return read_hash_map(rdr)
	default:
		return read_atom(rdr)
	}
	return read_atom(rdr)
}

func Read_str(str string) (MalType, error) {
	var tokens = tokenize(str)
	if len(tokens) == 0 {
		return nil, errors.New("<empty line>")
	}

	return read_form(&TokenReader{tokens: tokens, position: 0})
}

File: elixir/lib/mal/reader.ex
defmodule Mal.Reader do
  import Mal.Types

  def read_str(input) do
    case tokenize(input) do
      [] -> nil
      tokens -> tokens
        |> read_form
        |> elem(0)
    end
  end

  def tokenize(input) do
    regex = ~r/[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"|;.*|[^\s\[\]{}('"`,;)]*)/
    Regex.scan(regex, input, capture: :all_but_first)
      |> List.flatten
      |> List.delete_at(-1) # Remove the last match, which is an empty string
      |> Enum.filter(fn token -> not String.starts_with?(token, ";") end)
  end

  defp read_form([next | rest] = tokens) do
    case next do
      "(" -> read_list(tokens)
      "[" -> read_vector(tokens)
      "{" -> read_hash_map(tokens)
      "'" -> create_quote("quote", rest)
      "`" -> create_quote("quasiquote", rest)
      "~" -> create_quote("unquote", rest)
      "~@" -> create_quote("splice-unquote", rest)
      "@" -> create_quote("deref", rest)
      "^" -> create_meta(rest)
      ")" -> throw({:error, "unexpected )"})
      "]" -> throw({:error, "unexpected ]"})
      "}" -> throw({:error, "unexpected }"})
      _ ->
        token = read_atom(next)
        {token, rest}
    end
  end

  defp create_meta(tokens) do
    {meta, meta_rest} = read_form(tokens)
    {token, rest_tokens} = read_form(meta_rest)
    new_token = list([{:symbol, "with-meta"}, token, meta])
    {new_token, rest_tokens}
  end

  defp create_quote(quote_type, tokens) do
    {token, rest_tokens} = read_form(tokens)
    new_token = list([{:symbol, quote_type}, token])
    {new_token, rest_tokens}
  end

  defp read_list([_ | tokens]) do
    {ast, rest} = do_read_sequence(tokens, [], "(", ")")
    {list(ast), rest}
  end

  defp read_vector([_ | tokens]) do
    {ast, rest} = do_read_sequence(tokens, [], "[", "]")
    {vector(ast), rest}
  end

  defp read_hash_map([_ | tokens]) do
    {map, rest} = do_read_sequence(tokens, [], "{", "}")
    {hash_map(map), rest}
  end

  defp do_read_sequence([], _acc, _start_sep, end_sep), do: throw({:error, "expected #{end_sep}, got EOF"})
  defp do_read_sequence([head | tail] = tokens, acc, start_sep, end_sep) do
    cond do
      String.starts_with?(head, end_sep) ->
        {Enum.reverse(acc), tail}
      true ->
        {token, rest} = read_form(tokens)
        do_read_sequence(rest, [token | acc], start_sep, end_sep)
    end
  end

  defp read_atom("nil"), do: nil
  defp read_atom("true"), do: true
  defp read_atom("false"), do: false
  defp read_atom(":" <> rest), do: String.to_atom(rest)
  defp read_atom(token) do
    cond do
      String.starts_with?(token, "\"") and String.ends_with?(token, "\"") ->
        token
          |> String.slice(1..-2)
          |> String.replace("\\\"", "\"")

      integer?(token) ->
        Integer.parse(token)
          |> elem(0)

      true -> {:symbol, token}
    end
  end
end
